src/utils/Cypher.ts

```diff
@@ -26,19 +32,79 @@ export default class Cypher {
   }
 
-  public static encrypt(data: string): string {
+  public static async detectConfigurationKey(tenantID: string): Promise<void> {
+
+    const configCryptoKey: string = Configuration.getCryptoConfig().key;
+    const keySettings = await SettingStorage.getCryptoKeySettings(tenantID);
+    let cryptoKeySetting: CryptoKeySetting;
+
+    // Check if Crypto Key settings exists
+    if (keySettings) {
+      // Detect new config Cypher key
+      if (keySettings.cryptoKey.newKey && keySettings.cryptoKey.newKey !== configCryptoKey) {
+        cryptoKeySetting = {
+          oldKey: keySettings.cryptoKey.newKey
+        } as CryptoKeySetting;
+      } else if (!keySettings.cryptoKey.newKey && keySettings.cryptoKey.oldKey !== configCryptoKey) {
+        // If no newKey exist, detect oldKey change
+        cryptoKeySetting = {
+          oldKey: keySettings.cryptoKey.oldKey
+        } as CryptoKeySetting;
+      }
+      // If key change detected, prepare KeySettings change
+      if (cryptoKeySetting) {
+        await SettingStorage.deleteSetting(tenantID, keySettings.id);
+        cryptoKeySetting.newKey = configCryptoKey;
+
+        // Key migration of senitive data
+        this.cryptoKey = cryptoKeySetting;
+      }
+    } else {
+      // Create New Config Crypto Key in Tenant Settings
+      cryptoKeySetting = {
+        oldKey: configCryptoKey
+      } as CryptoKeySetting;
+    }
+
+    if (cryptoKeySetting) {
+      const keySettingToSave = {
+        identifier: TenantComponents.CRYPTO_KEY,
+        type: KeySettingsType.CRYPTO_KEY,
+        cryptoKey: cryptoKeySetting
+      } as KeySettings;
+      await SettingStorage.saveCryptoKeySettings(tenantID, keySettingToSave);
+    }
+  }
+
+  public static async migrateSensitiveData(tenantID: string, setting: SettingDB): Promise<void> {
+    if (this.cryptoKey) {
+      this.decryptSensitiveDataInJSON(setting, this.cryptoKey.oldKey);
+      this.encryptSensitiveDataInJSON(setting, this.cryptoKey.newKey);
+      await SettingStorage.saveSettings(tenantID, setting);
+    }
+  }
+
+  public static async migrateAllSensitiveData(tenantID: string, settings: SettingDB[]): Promise<void> {
+    for (const setting of settings) {
+      await this.migrateSensitiveData(tenantID, setting);
+    }
+  }
+
+  public static encrypt(data: string, key?: string): string {
     const iv = crypto.randomBytes(IV_LENGTH);
-    const cipher = crypto.createCipheriv(Cypher.getConfiguration().algorithm, Buffer.from(Cypher.getConfiguration().key), iv);
+    const cryptoKey = key ? key : Cypher.getConfiguration().key;
+    const cipher = crypto.createCipheriv(Cypher.getConfiguration().algorithm, Buffer.from(cryptoKey), iv);
     let encryptedData = cipher.update(data);
     encryptedData = Buffer.concat([encryptedData, cipher.final()]);
     return iv.toString('hex') + ':' + encryptedData.toString('hex');
   }
 
-  public static decrypt(data: string): string {
+  public static decrypt(data: string, key?: string): string {
```

Comment:
Same remark here.

Reply:
decrypt is also used in StripeBillingIntegration.ts and making key required throws error from there. Should I keep it optional and keep the logic for getting key from Config if not in CryptoSetting?